kie-issue#785: Integrate the Boxed Expression Editor into the new DMN Editor properties panel making it possible to edit its cells

ljmotta opened the pull request:
Closes: https://github.com/apache/incubator-kie-issues/issues/785

Description: Wip

jomarko commented:
Hello @ljmotta is available some deployed version of editor with your changes please? Would like to do some manual tests. Or do I need to build it locally on my own?

ljmotta commented:
@jomarko The PR is not ready for review yet. As soon it's ready for review I'll publish it.

jomarko commented:
### 02 Output Constraint
For Output header cells, we have a **Constraint** section, however only if the column is bound to a **Structure**? Is that expected? Because currently for columns bound to built -in data types, we do display **Constraint** section at all.

Maybe of options below would make the situation more 'clear':
- Do not display **Constraint** for any column, if that is currently never supported
- Display **Constraint** always, and say explicitly if that is supported or not for a given column
- maybe some other option

![Screenshot 2024-04-15 110321](https://github.com/apache/incubator-kie-tools/assets/8044780/7138f86c-6faa-4a9a-9a66-6f62df29b985)

jomarko commented:
### 03 Data Type Constraint and Monaco Suggestion
probably unrelated to the changes in the PR, however wanted to mention it here so we track it somehow.
Please notice adding data type constraint in combination of monaco suggestion causes an 'invalid Z index' issue? we see a button over suggestions.
![Screenshot 2024-04-15 111623](https://github.com/apache/incubator-kie-tools/assets/8044780/9c3120af-12f6-49d5-99f7-b6103095bd4b)

jomarko commented:
### 04 Commit the change
Imagine a situation like below, there is a cell selected and user put some description, **a** in our screenshot. Now imagine user want to proceed to the next cell, so he clicks to the next cell, the problem, maybe, maybe it is intended, is if user immediately click to another cell, then the **a** is committed into this second selected cell.

**Initial state**
![Screenshot 2024-04-15 114114](https://github.com/apache/incubator-kie-tools/assets/8044780/3c028360-e375-470e-af42-4b4082b0211c)

**Click to next cell**
![Screenshot 2024-04-15 114140](https://github.com/apache/incubator-kie-tools/assets/8044780/c6c7b5bf-5e6b-4e59-9d6d-9c3f4cacf05a)

jomarko commented:
### 05 Function parameters
Are in the properties panel written the same font as other text, maybe small change of font would make the properties panel more readable?

![Screenshot 2024-04-15 115808](https://github.com/apache/incubator-kie-tools/assets/8044780/d5ec3a6e-c5f4-408e-8d94-2f32088b98f2)

vs. using bold
![Screenshot 2024-04-15 121244](https://github.com/apache/incubator-kie-tools/assets/8044780/35402e51-812a-46b9-b47f-2f767f116ad0)

jomarko reviewed (commented):
### 07 Tests
Do we track somehow e2e playwright tests, when will we add them?

jomarko commented:
### 08 Back Compatibility
Opening an old model [1] has different result on legacy editor vs the new editor. please notice the decision table output header cell.

**Legacy Editor**
![Screenshot 2024-04-15 124051](https://github.com/apache/incubator-kie-tools/assets/8044780/15a41eec-0e91-4767-aa17-f16cb343ace5)

vs. **New Editor**
![Screenshot 2024-04-15 131001](https://github.com/apache/incubator-kie-tools/assets/8044780/8dc1ab8a-3c05-4721-b0e9-2160517924e0)



[1] https://raw.githubusercontent.com/redhat-cop/businessautomation-cop/master/testing-examples/cucumber-quarkus-example/src/main/resources/pricing.dmn

ljmotta commented:
@jomarko Thanks a lot for your review! I'll answer the raised observations inline:

1 - We have opted to use only constraints from custom data types, instead of letting create constraints on the Decision Table header. 
2 - As mentioned on 1, to create a constraint it’s necessary to create a custom data type, and not a `Structure`.
3 - It’s not related to this PR. I've confirmed on the `dailydev` env. Can you create a new issue so we can track this?
4 - This is a serious problem. I will take a look into it.
5 - I guess it would be good to incorporate this.
6 -
  - It shouldn’t be possible to change the node data type! Can you create an issue for this bug?
  - Regarding the boxed expression, we have disabled the option to open it.

7 -
  - We’re not tracking then, but I’ll add it to the 945 epic.
  - I’ll improve the comment. The disabled shortcuts are from Storybook and not from the Boxed Expression. It was causing a conflict with the Monaco editor.
  - The `description` is a `tag` of the DMN XSD. The `@_` is a property of a tag.

8 - This is a bug on the `dailydev` environment too. Can you create an issue to track this?

ljmotta commented:
@jomarko I've incorporated all your comments (4,5,7). Also, I've fixed all issues you've pointed out (3,6,8). The deployed page was updated. Please refresh the cookies to ensure you have the latest version (ctrl+shift+R)

jomarko commented:
Thank you for all updates and quick fix @ljmotta . I can confirm all points are fixed, however **08** seems still a little bit not ideal. When I open the model, in the output header of the decision table, the name is correct - **Best Proce** however it doesn't show **number** as data type, it still points to **Undefined**.
![Screenshot 2024-04-16 145526](https://github.com/apache/incubator-kie-tools/assets/8044780/70c37c46-6d44-4564-9b28-d51bdf9a55ea)

Is that expected somehow? I mean do older DMN versions specify DT output data type somehow incompatibly with DMN 1.5 spec so we can not derive it?

ljmotta commented:
@jomarko I overlooked the issue, sorry. I've already addressed and updated the deployed version.

jomarko reviewed (approved):
Thank you, @ljmotta! I think the original issue/comment **08** is solved very well now.

However, I found an interesting behavior, probably an issue with decision table top level output header cell. To me it seems to worth investigate in a separate ticket to not extend scope of this PR.

### DT-simple-01
![Screenshot 2024-04-18 195903](https://github.com/apache/incubator-kie-tools/assets/8044780/5c9773a2-a6df-4913-a2d4-71ab8e50d95a)

in this screenshot is an issue, that once node with DT expression is opened, the properties panel show the node data type, **Any** in the screenshot, but it is not synced with the boxed expression, see **<Undefined>** there.

### DT-simple-02
![Screenshot 2024-04-18 200015](https://github.com/apache/incubator-kie-tools/assets/8044780/f6a4b6aa-6985-4147-8e25-49fd3808e6e7)
In this screenshot is demoed, that the sync mechanism works well, if the change is done in boxed expression editor, we see the properties panel was synced. See the  **boolean**

However, what is strange in properties panel, we have all of these: **Decision Name**, **Decision Type**, **Column Name** and **Column Type**

### DT-complex-01
![Screenshot 2024-04-18 200100](https://github.com/apache/incubator-kie-tools/assets/8044780/b23ff1ea-d93e-4a45-911d-87ab017c2afd)
Is similar sync problem as DT-simple-01, again see **Any** type. 


The three listed issues/observations should be tracked/addressed separately form this PR to not extend the PR scope and make potential next PRs verification easier. I am approving this PR. Thank you, @ljmotta !

ljmotta commented:
@jomarko The `DT-simple-01` and `DT-complex-01` were a bug I've introduced on the previous commits, and it affected all expressions. Thanks for spotting it. Regarding `DT-simple-02`, when a Decision Table has only one output column, the Decision Table Output Header cell represents two cells. The Decision and the Column itself, thus we have the `Decision Name`, `Decision Type`, `Column Name` and `Column Type`. This becomes evident when we create a new output column, and now we will not have the `Decision Name`, `Decision Type`.

jomarko commented:
@ljmotta thank you for explanation of the `DT-simple-02`. I have rechecked `DT-simple-01` and `DT-complex-02` using [1] and the sync mechanism works without an issue.

@tiagobento thank you for reminder about new expression types. I can confirm using [1] the **For**, **Every** and **Some** cause an issue [2]. On the other hand, **If/Conditional** is working ok.


[1]
https://ljmotta.github.io/kie-tools-demos/kie-issues-785

[2]
![image](https://github.com/apache/incubator-kie-tools/assets/8044780/e186f08f-ebee-4c65-99dc-b1a380c9f9cc)

ljmotta commented:
@jomarko This issue was on main, I need to rebase and update my version. But I'll need to fix a few more things on the `idRandomizer` file.

jomarko commented:
I can still see the issue for: **For**, **Every** and **Some** expression. I guess there are still some pending updates on your site? Please let me know once I should recheck the behavior.

ljmotta commented:
@jomarko I've updated the deployed version. The last time you've tried it was missing a few adjustments. Now, it should work with the Iterator expressions.

ljmotta commented:
@jomarko I've finish with the changes, and I've re-deployed the PR.
